### PR TITLE
Validate certificate before sending back to proxy

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**
Recently we met errors from envoy 

`gRPC config for type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret rejected Failed to load certificate chain from <inline>`

This pr validates the fundamental format of the certs, and don't send invalid certs.

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [x] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Comment by howardjohn** [comment]
```
SecretGeneration/secrets-8                              48.8µs ± 3%       646.5µs ± 1%  +1224.85%  (p=0.008 n=5+5)
SecretGeneration/secrets#01-8                           5.10ms ± 3%       64.40ms ± 3%  +1161.76%  (p=0.008 n=5+5)
```

Pretty concerned about this

**Comment by hzxuzhonghu** [comment]
I guess mainly because the test read file again and again, i will make it read once.

**Comment by hzxuzhonghu** [comment]
Added env flag and make it false by default.

File: pilot/pkg/xds/sds.go [hunk 1]
```diff
@@ -134,39 +136,61 @@ func (s *SecretGen) Generate(proxy *model.Proxy, push *model.PushContext, w *mod
...
+	if isCAOnlySecret {
+		caCert, err := secretController.GetCaCert(sr.Name, sr.Namespace)
+		if err != nil {
+			pilotSDSCertificateErrors.Increment()
+			log.Warnf("failed to fetch ca certificate for %s: %v", sr.ResourceName, err)
+			return nil
+		}
+		if features.VerifySDSCertificate {
+			roots := x509.NewCertPool()
+			ok := roots.AppendCertsFromPEM(caCert)
```
Comment (howardjohn) [inline pilot/pkg/xds/sds.go:168]: should we just use `ParseCertificate` instead of a cert pool

File: pilot/pkg/xds/sds.go [hunk 2]
```diff
@@ -134,39 +136,61 @@ func (s *SecretGen) Generate(proxy *model.Proxy, push *model.PushContext, w *mod
...
-	return results, model.XdsLogDetails{AdditionalInfo: fmt.Sprintf("cached:%v/%v", cached, cached+regenerated)}, nil
+
+	key, cert, err := secretController.GetKeyAndCert(sr.Name, sr.Namespace)
+	if err != nil {
+		pilotSDSCertificateErrors.Increment()
+		log.Warnf("failed to fetch key and certificate for %s: %v", sr.ResourceName, err)
+		return nil
+	}
+	if features.VerifySDSCertificate {
+		if _, err := tls.X509KeyPair(cert, key); err != nil {
```
Comment (howardjohn) [inline pilot/pkg/xds/sds.go:186]: same `ParseCertificate` comment?

Comment (howardjohn) [reply pilot/pkg/xds/sds.go:186]: Well might need a little more I guess, there is a lot of validation in `X509KeyPair` and not a lot of waste. But I think CertPool is not very cheap

Comment (hzxuzhonghu) [reply pilot/pkg/xds/sds.go:186]: Will have a try and do a comparison

**Comment by hzxuzhonghu** [comment]
@howardjohn Much better now

SecretGeneration/secrets-8                              49.2µs ± 2%       217.0µs ± 2%  +341.40%  (p=0.008 n=5+5)
SecretGeneration/secrets#01-8                           5.14ms ± 3%       21.75ms ± 5%  +323.41%  (p=0.008 n=5+5)

**Comment by hzxuzhonghu** [comment]
ping @howardjohn

**Review by howardjohn — APPROVED** [review]

File: pilot/pkg/xds/sds.go [hunk 3]
```diff
@@ -134,39 +136,74 @@ func (s *SecretGen) Generate(proxy *model.Proxy, push *model.PushContext, w *mod
...
-	return results, model.XdsLogDetails{AdditionalInfo: fmt.Sprintf("cached:%v/%v", cached, cached+regenerated)}, nil
+
+	key, cert, err := secretController.GetKeyAndCert(sr.Name, sr.Namespace)
+	if err != nil {
+		pilotSDSCertificateErrors.Increment()
+		log.Warnf("failed to fetch key and certificate for %s: %v", sr.ResourceName, err)
+		return nil
+	}
+	if features.VerifySDSCertificate {
+		block, _ := pem.Decode(cert)
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/xds/sds.go:189]: make sense

**Pull request merged**